src/lib/tab_groups.ts

```diff
@@ -0,0 +1,330 @@
+ * Return a set of the current window's tab groups (empty if there are none).
+ *
+ */
+export async function tgroups() {
+    const groups = await browserBg.sessions.getWindowValue(await activeWindowId(), "tridactyl-tgroups")
+    return new Set<string>(groups as string[])
+}
+
+/**
+ * Set the current window's tab groups.
+ *
+ * @param groups The set of groups.
+ *
+ */
+export async function setTgroups(groups: Set<string>) {
+    return browserBg.sessions.setWindowValue(await activeWindowId(), "tridactyl-tgroups", [...groups])
+}
+
+/**
+ * Clear the current window's tab groups.
+ *
+ */
+export async function clearTgroups() {
+    return browserBg.sessions.removeWindowValue(await activeWindowId(), "tridactyl-tgroups")
+}
+
+/**
+ * Return the active tab group for the a window or undefined.
+ *
+ * @param id The id of the window. Use the current window if not specified.
+ *
+ */
+export async function windowTgroup(id?: number) {
+    return browserBg.sessions.getWindowValue(id || await activeWindowId(), "tridactyl-active-tgroup")
+}
+
+/**
+ * Set the active tab group for the current window.
+ *
+ * @param name The name of the new active tab group.
+ *
+ */
+export async function setWindowTgroup(name: string) {
+    return browserBg.sessions.setWindowValue(await activeWindowId(), "tridactyl-active-tgroup", name)
+}
+
+/**
+ * Clear the active tab group for the current window.
+ *
+ */
+export async function clearWindowTgroup() {
+    return browserBg.sessions.removeWindowValue(await activeWindowId(), "tridactyl-active-tgroup")
+}
+
+/**
+ * Return a tab's tab group.
+ *
+ * @param id The id of the tab. Use the current tab if not specified.
+ *
+ */
+export async function tabTgroup(id?: number) {
+    return browserBg.sessions.getTabValue(id || await activeTabId(), "tridactyl-tgroup")
+}
+
+/**
+ * Return a list of tab ids.
+ *
+ * @param id An id, a list of ids, or undefined.
+ *
+ * If id is undefined, return a list of the current tab id.
+ *
+ */
+async function tabIdsOrCurrent(ids?: number | number[]): Promise<number[]> {
+    if (!ids) {
+        ids = [await activeTabId()]
+    } else if (!Array.isArray(ids)) {
+        ids = [ids]
+    }
+    return ids
+}
+
+/**
+ * Set the a tab's tab group.
+ *
+ * @param name The name of the tab group.
+ * @param id The id(s) of the tab(s). Use the current tab if not specified.
+ *
+ */
+export async function setTabTgroup(name: string, id?: number | number[]) {
+    const ids = await tabIdsOrCurrent(id)
+    return Promise.all(ids.map(async (id) => {
+        await browserBg.sessions.setTabValue(id, "tridactyl-tgroup", name)
+    }));
+}
+
+/**
+ * Clear all the tab groups.
+ *
+ * @param id The id(s) of the tab(s). Use the current tab if not specified.
+ *
+ */
+export async function clearTabTgroup(id?: number | number[]) {
+    const ids = await tabIdsOrCurrent(id)
+    return Promise.all(ids.map(async (id) => {
+        await browserBg.sessions.removeTabValue(id, "tridactyl-tgroup")
+    }));
```

Comment:
Better to perform the removals in parallel:
```suggestion
    return Promise.all(ids.map(id => browserBg.sessions.removeTabValue(id, "tridactyl-tgroup")))
        .then(() => undefined);
```